Make sure you fill in any place that says `YOUR CODE HERE` or "YOUR ANSWER HERE", as well as your collaborators below:

In [1]:
COLLABORATORS = ""

---

In [2]:
import numpy as np

<div class="alert alert-warning"><b>N.B.: </b> This notebook has three sections that together make up half the problem set. Budget your time accordingly!</div>

The inhabitants of the planet Boton love pushing buttons. There are many kinds of buttons on Boton: red buttons, blue buttons, big buttons, small buttons. In fact, there are four dimensions along which buttons vary:

![](images/concepts.png)

Not all buttons do the same thing when pushed. Some are harmless, but others are dangerous and self-destruct. As young Botonans grow up, they are taught which buttons are safe to push and which are unsafe. Unfortunately, there are no hard-and-fast rules about which buttons are safe or unsafe, so young Botonans must develop a scheme for categorizing the buttons. For example, a Botonan might observe the following:
<a name="exemplars"></a>

![](images/exemplars.png)

Clearly, it's not enough to say that all the blue buttons are safe because there is at least one blue button known to be unsafe. So, how do the Botonans determine which buttons are safe?

---
## Context Model: Part A (0.5 points)

As a cognitive scientist, you know a few different models of categorization that have been proposed. For example, you might remember the *context model* described by Medin and Schaffer (1978). The context model is an exemplar model: the probability that a stimulus is assigned to a category is based on its similarity to all the exemplars in that category.

To use the context model, we represent each stimulus (button) as a vector. For example, we can represent a big square blue textured button as $\mathbf{x}=[1, 0, 1, 1]$. Similarly, we can represent the stimulus which is small red cicular textured button as $\mathbf{x}=[0, 1, 0, 1]$. Given this representation, we can now define a similarity function.

> <a name="eq:similarity"></a>The similarity $\mathbb{S}$ of one stimulus $\mathbf{x}$ to another stimulus $\mathbf{y}$ is given by the following set of equations:
>
> $$
\begin{align*}
\mathbb{S}(\mathbf{x}, \mathbf{y}) &= \prod_{i = 1}^m s(x_i, y_i)=s(x_1, y_1)\cdot{}s(x_2, y_2)\cdot{}\ldots{}\cdot{}s(x_m, y_m)\\
s(x_i, y_i) &= \left\{
\begin{array}{rl} 1 & \text{if } x_{i} = y_{i} \\
   \theta & \text{if } x_{i} \neq y_{i}\end{array}\right.
   \end{align*}
$$
>
> where $\theta$ is a constant.

(Remember that $\Pi$ is like $\Sigma$, except that you multiply the terms together instead of summing them. For example, $\prod_{k=1}^5{k} = 1 \cdot 2 \cdot 3 \cdot 4 \cdot 5 = 5! = 120$. Or, if $\mathbf{v}=[4,1,2]$ then, $\prod_{i=1}^3{v_i}=4 \cdot 1 \cdot 2 = 8$.)

So, if one stimulus is $\mathbf{x}=[1, 0, 1, 1]$ and the other is $\mathbf{y}=[0, 1, 1, 0]$, then we have:

$$
\begin{align*}
\mathbb{S}(\mathbf{x}, \mathbf{y}) &= \prod_{i=1}^4 s(x_i, y_i) \\
&= s(x_1, y_1) \cdot{} s(x_2, y_2) \cdot{} s(x_3, y_3) \cdot{} s(x_4, y_4) \\
&= \theta \cdot{} \theta \cdot{} 1 \cdot{} \theta \\
&= \theta^3
\end{align*}
$$

<div class="alert alert-success">Complete the function `calculate_similarity` to calculate the similarity $\mathbb{S}(\mathbf{x}, \mathbf{y})$ between two stimuli (as defined in the [similarity equation](#eq:similarity)).</div>

In [3]:
def calculate_similarity(x, y, theta=0.1):
    """Calculates the similarity between a stimulus x and a 
    stimulus y, where similarity is defined as:
    
        S(x, y) = s(x_1, y_1) * s(x_2, y_2) * ... * s(x_m, y_m)
    
    and:
    
        s(x_i, y_i) = 1 if x_i == y_i, theta otherwise
        
    Note: your solution can be done in one line of code, including 
    the return statement. Think about how many times you need to 
    multiply theta with itself. How can you easily compute this 
    number?
    
    Parameters
    ----------
    x, y : numpy arrays with shape (m,)
        The stimuli to compute similarity between
    theta : (optional) float
        A parameter to the similarity function. When the function is
        called without theta having been specified, it defaults to 0.1.
        
    Returns
    -------
    float : the similarity between x and y
    
    """
    n = 0
    sim = []
    while n < len(x):
        if x[n] == y[n]:
            sim.append(1)
        else:
            sim.append(theta)
        n = n+1
    return np.prod(sim)

Verify that your function works on the example we worked out above (is the answer it returns equivalent to $\theta^3$?)

In [4]:
x = np.array([1, 0, 1, 1])
y = np.array([0, 1, 1, 0])
print("S(x, y) = {}".format(calculate_similarity(x, y)))
print("S(x, y, theta=0.3) = {}".format(calculate_similarity(x, y, theta=0.3)))

S(x, y) = 0.0010000000000000002
S(x, y, theta=0.3) = 0.027


In [5]:
# add your own test cases here!


In [6]:
"""Check that calculate_similarity works correctly."""

from numpy.testing import assert_allclose

x = np.ones(8)
y = np.zeros(8)
assert_allclose(calculate_similarity(x, x), 1.0)
assert_allclose(calculate_similarity(x, y), 1e-08)
assert_allclose(calculate_similarity(x, y, theta=2.0), 256.0)
assert_allclose(calculate_similarity(x, y, theta=0.47), 0.002381128666176099)
assert_allclose(calculate_similarity(x, y, theta=1), 1)
assert_allclose(calculate_similarity(x, y, theta=1.2), 4.2998169599999985)

x = np.array([1, 0, 1, 0])
y = np.array([0, 0, 1, 1])
assert_allclose(calculate_similarity(x, y), 0.01)
assert_allclose(calculate_similarity(x, y, theta=0.2), 0.04)
assert_allclose(calculate_similarity(x, y, theta=0.3), 0.09)
assert_allclose(calculate_similarity(x, y, theta=0.47), 0.22089999999999999)
assert_allclose(calculate_similarity(x, y, theta=1), 1)
assert_allclose(calculate_similarity(x, y, theta=1.2), 1.44)

x = np.array([0, 1, 1, 1, 0, 0, 0, 0, 0, 1])
y = np.array([1, 1, 1, 0, 0, 1, 0, 1, 1, 0])
assert_allclose(calculate_similarity(x, y), 1e-6)
assert_allclose(calculate_similarity(x, y, theta=0.2), 6.4e-5)
assert_allclose(calculate_similarity(x, y, theta=0.3), 0.0007289999999999998)
assert_allclose(calculate_similarity(x, y, theta=0.47), 0.010779215328999996)
assert_allclose(calculate_similarity(x, y, theta=1), 1)
assert_allclose(calculate_similarity(x, y, theta=1.2), 2.9859839999999993)

print("Success!")

Success!


---

## Context Model: Part B (1.75 points)

Now that we have defined the similarity between two stimuli, we can take a look at the equation for the context model. In the definition below, you can think of category $A$ as being the *safe buttons*, and category $B$ as being the *unsafe buttons*.

> <a name="eq:context-model"></a>The context model, which gives the probability that a novel stimulus $\mathbf{x}$ belongs to category $A$ (as opposed to category $B$) is given by:
>
> $$
P(A|\mathbf{x}) = \frac{\sum_{\mathbf{a} \in A} \mathbb{S}(\mathbf{x}, \mathbf{a})}{\sum_{\mathbf{a} \in A} \mathbb{S}(\mathbf{x}, \mathbf{a})+ \sum_{\mathbf{b} \in B} \mathbb{S}(\mathbf{x}, \mathbf{b})}
$$
>
> where $\sum_{\mathbf{a} \in A} \mathbb{S}(\mathbf{x}, \mathbf{a})$ is the sum over the similarity of $\mathbf{x}$ to all exemplars $\mathbf{a}$ in category $A$, and $\sum_{\mathbf{b} \in B} \mathbb{S}(\mathbf{x}, \mathbf{b})$ is the sum over the similarity of $\mathbf{x}$ to all exemplars $\mathbf{b}$ in category $B$.

Note that because $P(A|\mathbf{x})$ is a probability, we can easily compute from it the probability that $\mathbf{x}$ belongs to category $B$. The stimulus *must* belong to one of the two categories, thus $P(B|\mathbf{x})=1-P(A|\mathbf{x})$.

Let's work through an example of this. Suppose there is one test stimulus $\mathbf{x} = [1, 0, 1]$, and four exemplars: $\mathbf{a}_1=[0, 0, 0]$ and $\mathbf{a}_2=[0, 0, 1]$ (which have category labels of $A$) and $\mathbf{b}_1=[1, 1, 1]$ and $\mathbf{b}_2=[0, 1, 1]$ (which have category labels of $B$). Then:

$$
P(A|\mathbf{x}) = \frac{\sum_{\mathbf{a}\in A} \mathbb{S}(\mathbf{x}, \mathbf{a})}{\sum_{\mathbf{a}\in A} \mathbb{S}(\mathbf{x}, \mathbf{a}) + \sum_{\mathbf{b}\in B} \mathbb{S}(\mathbf{x}, \mathbf{b})}=\frac{\mathbb{S}(\mathbf{x}, \mathbf{a}_1) + \mathbb{S}(\mathbf{x}, \mathbf{a}_2)}{\mathbb{S}(\mathbf{x}, \mathbf{a}_1) + \mathbb{S}(\mathbf{x}, \mathbf{a}_2) + \mathbb{S}(\mathbf{x}, \mathbf{b}_1) + \mathbb{S}(\mathbf{x}, \mathbf{b}_2)}
$$

where $\mathbb{S}$ is the [similarity function](#eq:similarity) that we defined above and implemented in `calculate_similarity`.

<div class="alert alert-success">Write code that uses your function `calculate_similarity` to implement a version of the context model (i.e., computes the [context model equation](#eq:context-model)). Complete the function `context_model` with your solution.</div>

In [7]:
def context_model(test_stimuli, exemplars, exemplar_categories, theta=0.1):
    """Computes the probability that each test stimulus belongs to 
    category A.
    
    Note: your solution can be done in 7 lines of code, including 
    the return statement
    
    Parameters
    ----------
    test_stimuli : numpy array with shape (n, m)
        n stimuli, each with m features, to be classified (i.e. 
        compute P(A|x) for each x)
    exemplars : numpy array with shape (k, m)
        k exemplars, each with m features
    exemplar_categories : numpy string array with shape (k,)
        Categories for the k exemplars. You can assume the values of 
        exemplar_categories will always be either be "A" or "B".
    theta : (optional) float
        A parameter to pass to the similarity function.
        
    Returns
    -------
    numpy array with shape (n,) such that the i^th element 
    corresponds to P(A|test_stimuli[i])
        
    """
    a_exemplars, b_exemplars = [], []
    a1, b1 = [], []
    n = 0
    final = []
    for ex in exemplar_categories:
        if ex == "A":
            a_exemplars.append(exemplars[n])
            n = n+1
        if ex == "B":
            b_exemplars.append(exemplars[n])
            n = n+1
    for stim in test_stimuli:
        for a_ex in a_exemplars:
            a1.append(calculate_similarity(stim, a_ex, theta))
        for b_ex in b_exemplars:
            b1.append(calculate_similarity(stim, b_ex, theta))
        final.append(sum(a1) / (sum(a1) + sum(b1)))
        a1.clear()
        b1.clear()
    return np.array(final)

In [8]:
# add your own test cases here!


In [9]:
"""Check that the context model is correct."""

# check that output has the correct shape and is always probabilties
for i in range(100):
    n, m, k = np.random.randint(2, 10, size=(3,))
    test_stimuli = np.random.choice(np.array([1, 0]), size=(n,m))
    test_exemplars = np.random.choice(np.array([1, 0]), size=(k,m))
    test_exemplar_categories = np.append(np.random.choice(["A", "B"], size=(k-2,)), ["A", "B"])
    ps = context_model(test_stimuli, test_exemplars, test_exemplar_categories)
    
    assert ps.shape == (n,)
    assert ((ps >= 0) & (ps <= 1)).all()

# check that output is exactly right for a small example
test_stimuli = np.array([[1, 0, 1], [0, 1, 1], [1, 1, 1]])
test_exemplars = np.array([[1, 1, 0], [0, 0, 1], [0, 0, 0], [1, 0, 0]])
test_exemplar_categories = np.array(["B", "A", "A", "A"])
assert_allclose(
    context_model(test_stimuli, test_exemplars, test_exemplar_categories),
    np.array([ 0.95454545,  0.91735537,  0.17355372]))
assert_allclose(
    context_model(test_stimuli, test_exemplars, test_exemplar_categories, theta=0.2),
    np.array([ 0.91666667,  0.86111111,  0.30555556]))

# check that output is exactly right for a larger example
test_stimuli = np.array([
    [1, 1, 1, 1, 0],
    [0, 0, 1, 1, 0],
    [1, 0, 0, 1, 1],
    [1, 1, 0, 1, 1]])
test_exemplars = np.array([
    [0, 0, 1, 0, 1],
    [0, 1, 1, 1, 0],
    [1, 0, 0, 0, 1],
    [0, 0, 1, 1, 1],
    [0, 0, 1, 0, 0],
    [1, 0, 0, 1, 1]])
test_exemplar_categories = np.array(['B', 'A', 'B', 'B', 'B', 'A'])
assert_allclose(
    context_model(test_stimuli, test_exemplars, test_exemplar_categories),
    np.array([ 0.97868217,  0.32465445,  0.900018  ,  0.9009009 ]))
assert_allclose(
    context_model(test_stimuli, test_exemplars, test_exemplar_categories, theta=0.5),
    np.array([ 0.625     ,  0.32258065,  0.53125   ,  0.57142857]))

# check that it uses calculate_similarity
old_calculate_similarity = calculate_similarity
del calculate_similarity
try:
    context_model(test_stimuli, test_exemplars, test_exemplar_categories)
except NameError:
    pass
else:
    raise AssertionError("context_model does not call the calculate_similarity function")
finally:
    calculate_similarity = old_calculate_similarity
    del old_calculate_similarity
    
print("Success!")

Success!


---

## Context Model: Part C (0.5 points)

Now that you have an implementation of the context model, let's try it out! First, let's see how well it does at categorizing the [exemplars that we already have](#exemplars):

In [10]:
safe_exemplars = np.array([
    [0, 1, 1, 0], # circle / small / blue / solid
    [1, 0, 1, 0], # square / big   / blue / solid
    [1, 1, 0, 0], # square / small / red  / solid
    [1, 1, 1, 1]  # square / small / blue / textured
])
unsafe_exemplars = np.array([
    [1, 1, 0, 1], # square / small / red  / textured
    [0, 0, 0, 1], # circle / big   / red  / textured
    [0, 1, 1, 1], # circle / small / blue / textured
    [0, 1, 0, 0]  # circle / small / red  / solid
])
all_exemplars = np.vstack([safe_exemplars, unsafe_exemplars])
exemplar_categories = np.array(["A", "A", "A", "A", "B", "B", "B", "B"])

In [11]:
context_model(all_exemplars, all_exemplars, exemplar_categories)

array([ 0.83603896,  0.99613153,  0.83603896,  0.83603896,  0.16396104,
        0.00386847,  0.16396104,  0.16396104])

<div class="alert alert-success">Describe how well the context model does at categorizing the known exemplars. Does it give a higher value for $P(A|\mathbf{x})$ for those $\mathbf{x}$ which are actually in category $A$? (**0.5 points**)</div>

The context model does a very good job at categorizing the exemplars. It gives a much higher probability value for the x's that are in category A (the safe ones), and gives a much smaller probability value for the 'unsafe' cateogory B exemplars. Using this, the context model could very accurately categorize between safe and unsafe buttons.

---

## Context Model: Part D (0.5 points)

A young Botonan is strolling through the city, when a flash of blue and red is seen coming from an alleyway. Taking a closer look, the youngster discovers four buttons never before encountered:
<a name="novel"></a>

![](images/novel.png)

As a Botonan, this youngster feels a strong urge to push the buttons. However, the possibility that some of the buttons might be dangerous demands restraint. The Botonan pauses and takes a closer look.

Let's see what the context model says about how likely these buttons are to be dangerous:

In [12]:
novel_stimuli = np.array([
    [0, 0, 0, 0], # circle / big   / red  / solid
    [1, 0, 0, 1], # square / big   / red  / textured
    [1, 0, 1, 1], # square / big   / blue / textured
    [0, 0, 1, 0]  # circle / big   / blue / solid
])

In [13]:
context_model(novel_stimuli, all_exemplars, exemplar_categories)

array([ 0.12968548,  0.12968548,  0.87031452,  0.87031452])

<div class="alert alert-success">In words, what does the context model say? Which of these buttons are more likely to belong to category A (safe buttons)? (**0.25 points**)</div>

The context model says that the first two red buttons are unsafe (and should belong to category B), and the second two blue buttons are safe (and should belong to category A). This is because the first two have a low probability of being safe $P(A|\mathbf{x}) = 0.12 $, and the second two have a higher probability of being safe $P(A|\mathbf{x}) = 0.87 $.

<div class="alert alert-success">If the context model is an accurate model of how Botonans categorize concepts, do you think they would push any of the buttons? (**0.25 points**)</div>

Yes, they would likely push the two blue buttons because the probability that they are safe is around the range of other known safe buttons in category A.

---
## Prototype Model: Part A (0.5 points)

You also know of another model of categorization: the prototype model. This model is similar to the context model, but rather than comparing the new stimulus to all of the exemplars, it compares the new stimulus to the category *prototypes*. How do we know what the category prototype is? Recall from the last problem set that we can compute a prototype from a set of exemplars.

<div class="alert alert-success">Complete the following `prototype` function (it is fine if you reuse your code from the last problem set--- but make sure that it actually works as expected!).</div>

In [14]:
def prototype(features):
    """
    Compute the prototype features, based on the given features of
    category members. The prototype should have a feature if half or
    more of the category members have that feature.

    Hint: you can use your implementation of `prototype` from the
    last problem set here, though you may want to cast the array to 
    integers to match the format elsewhere (try `.astype(int)`).
    
    Your solution can be done in 1 line of code (including the return 
    statement).
    
    Parameters
    ----------
    features : numpy array with shape (n, m)
        The first dimension corresponds to n category members, and the
        second dimension to m features.
    
    Returns
    -------
    numpy array with shape (m,) corresponding to the features
    of the prototype of the category members
    
    """
    #reuse code from Ps2
    proto = []
    lst = []
    proto.append(np.mean(features, axis = 0))
    for p in proto[0]:
        if p >= 0.5:
            lst.append(True)
        else:
            lst.append(False)
    return np.array(lst).astype(int)

In [15]:
# add your own test cases here!


In [16]:
"""Test the prototype function."""
from nose.tools import assert_equal
from numpy.testing import assert_array_equal

# make sure they get features that are half or more
assert_array_equal(prototype(np.array([[0, 1], [0, 0]])), np.array([0, 1]))

for i in range(10):
    # create a random array of features
    n, m = np.random.randint(10, 100, 2)
    features = np.random.randint(0, 2, (n, m))
    
    # compute the prototype
    proto = prototype(features)
    
    # check the shape and type
    assert_equal(proto.shape, (m,), "incorrect shape for the prototype array")
    
    # check that the prototype is correct
    for j in range(m):
        count = features[:, j].sum()
        if count >= (n / 2) and not proto[j]:
            raise AssertionError("prototype should have feature {}, but it doesn't".format(j))
        elif count < (n / 2) and proto[j]:
            raise AssertionError("prototype should NOT have feature {}, but it does".format(j))

print("Success!")

Success!


<a name="prototypes"></a>Check what your function returns for the "safe button" prototype and the "unsafe button" prototype:

In [17]:
safe_prototype = prototype(safe_exemplars)
unsafe_prototype = prototype(unsafe_exemplars)
print("Safe button prototype: {}".format(safe_prototype))
print("Unsafe button prototype: {}".format(unsafe_prototype))

Safe button prototype: [1 1 1 0]
Unsafe button prototype: [0 1 0 1]


<div class="alert alert-success">Describe in words what the features of each prototype are (e.g., "the safe button prototype is a [size], [color], [fill] [shape]"). Remember that the first feature corresponds to shape, the second feature corresponds to size, the third feature corresponds to color, and the fourth feature corresponds to fill. (**0.25 points**)</div>

The safe button prototype corresponds to: square, small, blue, solid. The unsafe button prototype corresponds to: circle, small, red, textured. 

---

## Prototype Model: Part B (1.75 points)

Now that we have a way of computing prototypes from our exemplars, let's take a look at the actual prototype model.

<a name="eq:prototype"></a>
> The prototype model can be described by the following equation:
>
> $$
P(A|\mathbf{x})=\frac{\mathbb{S}(\mathbf{x}, \mathbf{\mu}_A)}{\mathbb{S}(\mathbf{x}, \mathbf{\mu}_A) + \mathbb{S}(\mathbf{x}, \mathbf{\mu}_B)}
$$
>
> where $\mathbb{S}$ is the [similarity function defined above](#eq:similarity), $\mathbf{x}$ is the novel stimulus, $\mathbf{\mu}_A$ is the prototype of category $A$, and $\mathbf{\mu}_B$ is the prototype of category $B$.

<div class="alert alert-success">Using your `calculate_similarity` function and your `prototype` function, complete the function `prototype_model` below to implement the [prototype model equation](#eq:prototype).</div>

In [18]:
def prototype_model(test_stimuli, exemplars, exemplar_categories, theta=0.1):
    """Computes the probability that each test stimulus belongs to 
    category A.
    
    Note: your solution can be done in 8 lines of code, including
    the return statement.

    Parameters
    ----------
    test_stimuli : numpy array with shape (n, m)
        n stimuli, each with m features, to be classified (i.e. 
        compute P(A|x) for each x)
    exemplars : numpy array with shape (k, m)
        k exemplars, each with m features
    exemplar_categories : numpy string array with shape (k,)
        Categories for the k exemplars. You can assume the values of 
        exemplar_categories will always be either be "A" or "B".
    theta : (optional) float
        A parameter to pass to the similarity function. If theta is not
        specified, it defaults to 0.1.
        
    Returns
    -------
    numpy array with shape (n,) such that the i^th element 
    corresponds to P(A|test_stimuli[i])
        
    """
    prob = []
    a_exemplars, b_exemplars = [], []
    n = 0
    for ex in exemplar_categories:
        if ex == "A":
            a_exemplars.append(exemplars[n])
            n = n+1
        if ex == "B":
            b_exemplars.append(exemplars[n])
            n = n+1
    proto_a = prototype(a_exemplars)
    proto_b = prototype(b_exemplars)
    for stim in test_stimuli:
        sim_a = calculate_similarity(stim, proto_a, theta)
        sim_b = calculate_similarity(stim, proto_b, theta)
        prob.append(sim_a / (sim_a + sim_b))
    return np.array(prob)

In [19]:
# add your own test cases here!


In [20]:
"""Check that the prototype model is correct."""

# check that output has the correct shape and is always probabilties
for i in range(100):
    n, m, k = np.random.randint(2, 10, size=(3,))
    test_stimuli = np.random.choice(np.array([1, 0]), size=(n,m))
    test_exemplars = np.random.choice(np.array([1, 0]), size=(k,m))
    test_exemplar_categories = np.append(np.random.choice(["A", "B"], size=(k-2,)), ["A", "B"])
    ps = prototype_model(test_stimuli, test_exemplars, test_exemplar_categories)
    
    assert ps.shape == (n,)
    assert ((ps >= 0) & (ps <= 1)).all()

# check that output is exactly right for a small example
test_stimuli = np.array([[1, 0, 1], [0, 1, 1], [1, 1, 1]])
test_exemplars = np.array([[1, 1, 0], [0, 0, 1], [0, 0, 0], [1, 0, 0]])
test_exemplar_categories = np.array(["B", "A", "A", "A"])
assert_allclose(
    prototype_model(test_stimuli, test_exemplars, test_exemplar_categories),
    np.array([ 0.5       ,  0.5       ,  0.00990099]))
assert_allclose(
    prototype_model(test_stimuli, test_exemplars, test_exemplar_categories, theta=0.2),
    np.array([ 0.5       ,  0.5       ,  0.03846154]))

# check that output is exactly right for a larger example
test_stimuli = np.array([
    [1, 1, 1, 1, 0],
    [0, 0, 1, 1, 0],
    [1, 0, 0, 1, 1],
    [1, 1, 0, 1, 1]])
test_exemplars = np.array([
    [0, 0, 1, 0, 1],
    [0, 1, 1, 1, 0],
    [1, 0, 0, 0, 1],
    [0, 0, 1, 1, 1],
    [0, 0, 1, 0, 0],
    [1, 0, 0, 1, 1]])
test_exemplar_categories = np.array(['B', 'A', 'B', 'B', 'B', 'A'])
assert_allclose(
    prototype_model(test_stimuli, test_exemplars, test_exemplar_categories),
    np.array([ 0.999001  ,  0.09090909,  0.90909091,  0.999001  ]))
assert_allclose(
    prototype_model(test_stimuli, test_exemplars, test_exemplar_categories, theta=0.5),
    np.array([ 0.88888889,  0.33333333,  0.66666667,  0.88888889]))

# check that it uses calculate_similarity
old_calculate_similarity = calculate_similarity
del calculate_similarity
try:
    prototype_model(test_stimuli, test_exemplars, test_exemplar_categories)
except NameError:
    pass
else:
    raise AssertionError("prototype_model does not call the calculate_similarity function")
finally:
    calculate_similarity = old_calculate_similarity
    del old_calculate_similarity
    
# check that it uses prototype
old_prototype = prototype
del prototype
try:
    prototype_model(test_stimuli, test_exemplars, test_exemplar_categories)
except NameError:
    pass
else:
    raise AssertionError("prototype_model does not call the prototype function")
finally:
    prototype = old_prototype
    del old_prototype
    
print("Success!")

Success!


---

## Prototype Model: Part C (0.5 points)

Let's see how well the prototype model does at categorizing the prototypes of safe and unsafe buttons:

In [21]:
prototypes = np.vstack([safe_prototype, unsafe_prototype])
prototype_model(prototypes, all_exemplars, exemplar_categories)

array([ 0.999001,  0.000999])

<div class="alert alert-success">How well does it do? Does it classify the safe prototype as safe, and the unsafe prototype as unsafe? (**0.5 points**)</div>

The prototype does extremely well at classifying the safe prototypes as safe and the unsafe prototypes as unsafe. It classsifies the safe buttons being in category A with a probability of 0.999, and the unsafe buttons being in category A with 0.0009 probability (they are therefore category B / unsafe buttons). 

---

## Prototype Model: Part D (0.5 points)

Now let's see what the prototype model would say about the [unknown buttons](#novel) that our young Botonan has encountered.

In [22]:
prototype_model(novel_stimuli, all_exemplars, exemplar_categories)

array([ 0.09090909,  0.09090909,  0.90909091,  0.90909091])

<div class="alert alert-success">What does the prototype model say? Which of these buttons are more likely to belong to category A (safe buttons)? (**0.25 points**)</div>

It says the first two red buttons are extremely unlikely to belong to category A, because $P(A|\mathbf{x}) = 0.0909$ for the two buttons, and are very likely to be unsafe. Therefore, the second two blue buttons (both with $P(A|\mathbf{x}) = 0.9090$) are much more likely to belong to category A, and thus must be the safe buttons.

<div class="alert alert-success">What does the prototype model say? Which of these buttons are more likely to belong to category A (safe buttons)? If the prototype model is an accurate model of how Botonans categorize concepts, do you think they would push any of the buttons? (**0.25 points**)</div>

The prototype model says that the first two buttons are unsafe to push, and the second two are highly likely to be safe. The Botonans would likely push the second two because there is a 90% chance they will be correct.

---
## Comparing Models: Part A (0.75 points)

As we develop computational models of cognition, it is not enough to look at a single model and declare that it is good or bad. What is it good or bad in relation to? We must always *compare* our models, in order to get a better sense of how the space of models behave on a particular type of problem.

We have now analyzed the context model and prototype model independently, but we have not compared them. First, let's compare how they both behave on the exemplars that have already been observed:

In [23]:
print("Context model on exemplars:    {}".format(context_model(all_exemplars, all_exemplars, exemplar_categories)))
print("Prototype model on exemplars:  {}".format(prototype_model(all_exemplars, all_exemplars, exemplar_categories)))
print("True exemplar categories:      {}".format(exemplar_categories))

Context model on exemplars:    [ 0.83603896  0.99613153  0.83603896  0.83603896  0.16396104  0.00386847
  0.16396104  0.16396104]
Prototype model on exemplars:  [ 0.90909091  0.999001    0.90909091  0.90909091  0.09090909  0.000999
  0.09090909  0.09090909]
True exemplar categories:      ['A' 'A' 'A' 'A' 'B' 'B' 'B' 'B']


<div class="alert alert-success">Which model is more accurate at predicting the exemplar categories? (**0.25 points**)</div>

The prototype model is more accurate at prediciting the exemplar categories because the probability is higher than the context model when it is in category A, as well as lower when it is in category B. 

<div class="alert alert-success">Even though the models have seen these exemplars, and know their true categories, they do not predict the category labels with 100% certainty. Why is this the case? (**0.5 points**)</div>

These models do not predict the category labels with 100% certainty because for the context model, they must compare them to all of the examplars that belong in each category -- and even if the exemplar is in the category, it is still compared to other ones -- which may have different features than this one, therfore never being able to reach 100% certainty. For the prototype model, the model is comparing the exemplar to the prototype of the category, and thus may have a few matching features but not all, because the prototype is based off if the features are in 50% of the exemplars. 

---

## Comparing Models: Part B (0.75 points)

Now let's take a look at how well the models perform on the prototypes:

In [24]:
print("Context model on prototypes:    {}".format(context_model(prototypes, all_exemplars, exemplar_categories)))
print("Prototype model on prototypes:  {}".format(prototype_model(prototypes, all_exemplars, exemplar_categories)))
print("True prototype categories:      {}".format(np.array(["A", "B"])))

Context model on prototypes:    [ 0.93001628  0.06998372]
Prototype model on prototypes:  [ 0.999001  0.000999]
True prototype categories:      ['A' 'B']


<div class="alert alert-success">Which model is more accurate at predicting the prototypes? Why? (**0.75 points**)</div>

The Prototype model is more accurate at predicting the prototypes. This is because in the probability function of this model, it calculates the similarity of each novel stimulus to the prototype of categories A and B. Therfore, it emits a much more accurate way of predicting the prototypes.

---
## Comparing Models: Part C (1 point)

Finally, let's take a look at how the models predict the novel stimuli:

In [25]:
print("Context model on novel:   {}".format(context_model(novel_stimuli, all_exemplars, exemplar_categories)))
print("Prototype model on novel: {}".format(prototype_model(novel_stimuli, all_exemplars, exemplar_categories)))

Context model on novel:   [ 0.12968548  0.12968548  0.87031452  0.87031452]
Prototype model on novel: [ 0.09090909  0.09090909  0.90909091  0.90909091]


<div class="alert alert-success">If the true categories of the novel stimuli were $[B, B, B, A]$, which model would be more advantageous in this scenario (remember that category $B$ is the unsafe buttons)? Explain your answer. (**0.5 points**)</div>

If the true categories were $[B, B, B, A]$, I believe the context model would be more advantageous than the Prototype model. This is because on the third novel simulus, since the true category is B, the context model predicts it will be less likeley to be in category A than in B, so it predicts it to be in category with around a 3% more accruate evaluation than the prototype model. Therefore, since it predicts the third novel stimuli to be more unsafe than the prototype model, it would be more advantageous.

<div class="alert alert-success">If the true categories of the novel stimuli were $[B, B, A, A]$? Explain your answer. (**0.5 points**)</div>

If the true categories were $[B, B, A, A]$, the prototype model would be more advantageous than the context model because it ranks the first two (unsafe) buttons with lower probability (3% less) that they would be in category A, meaning they are more unsafe (category B), which corresponds to the first two true categories (B,B). In addition, it ranks the last two (safe) buttons with a higher probability (3% more) than the context model that they would be safe (category A), which corresponds to the last two true categories (A,A).

---
## Comparing Models: Part D (1 point)

<div class="alert alert-success"> Participants in experiments are better (more accurate and/or faster) at classifying a prototype of previously seen stimuli than classifying new stimuli generated from that prototype: an effect called the **prototype effect**. Would the context / exemplar model be able to account for this effect in people? Why is it that the context / exemplar model can/cannot explain this effect? (**0.5 points**)</div>

The context model would not be able to account for this effect because it would be slower due to the process of comparing the new exemplar to all the exemplars in the category. 

<div class="alert alert-success"> Would the prototype model be able to account for this effect in people? Why is it that the prototype model can/cannot explain this effect? (**0.5 points**)</div>

Yes, the prototype model would be able to account for the prototype effect because it is comparing the new exemplar to the prototype which it has already created.

---

Before turning this problem in remember to do the following steps:

1. **Restart the kernel** (Kernel$\rightarrow$Restart)
2. **Run all cells** (Cell$\rightarrow$Run All)
3. **Save** (File$\rightarrow$Save and Checkpoint)

<div class="alert alert-danger">After you have completed these three steps, ensure that the following cell has printed "No errors". If it has <b>not</b> printed "No errors", then your code has a bug in it and has thrown an error! Make sure you fix this error before turning in your problem set.</div>

In [26]:
print("No errors!")

No errors!
